test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,327 @@
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	crdv1alpha1 "antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	"antrea.io/antrea/pkg/features"
+	. "antrea.io/antrea/test/e2e/utils"
+)
+
+const (
+	interfaceNameLength = 5
+	nameSpace           = "vm-ns"
+	serviceAccount      = "vm-agent"
+)
+
+type vmInfo struct {
+	nodeName string
+	osType   string
+	ifName   string
+	ip       string
+	ifIndex  string // Used only for Windows
+}
+
+// TestVMAgent is the top-level test which can contain some subtests for
+// VMAgent so they can share setup, teardown.
+func TestVMAgent(t *testing.T) {
+	skipIfFeatureDisabled(t, features.ExternalNode, false, true)
+	skipIfNoVMs(t)
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+	vmList, err := setupVMAgentTest(t, data)
+	if err != nil {
+		t.Fatalf("Error when setting up VMAgent test: %v", err)
+	}
+	defer teardownVMAgentTest(t, data, vmList)
+	t.Run("testExternalNode", func(t *testing.T) { testExternalNode(t, data, vmList) })
+}
+
+// setupVMAgentTest creates ExternalNode, starts antrea-agent
+// and returns a list of VMs upon success
+func setupVMAgentTest(t *testing.T, data *TestData) ([]vmInfo, error) {
+	t.Logf("List of Windows VMs: '%s', Linux VMs: '%s'", testOptions.windowsVMs, testOptions.linuxVMs)
+	nsReturned, _ := data.clientset.CoreV1().Namespaces().Get(context.TODO(), nameSpace, metav1.GetOptions{})
+	saReturned, _ := data.clientset.CoreV1().ServiceAccounts(nameSpace).Get(context.TODO(), serviceAccount, metav1.GetOptions{})
+	t.Logf("Using service account %s", saReturned.Name)
+	t.Logf("Using namespace %s", nsReturned.Name)
+	var vmList []vmInfo
+	if testOptions.linuxVMs != "" {
+		vms := strings.Split(testOptions.linuxVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Linux VM: %s", vm)
+			tempVm := getVMInfo(t, data, vm)
+			vmList = append(vmList, tempVm)
+		}
+	}
+	if testOptions.windowsVMs != "" {
+		vms := strings.Split(testOptions.windowsVMs, ",")
+		for _, vm := range vms {
+			t.Logf("Get info for Windows VM: %s", vm)
+			tempVm := getWindowsVMInfo(t, data, vm)
+			vmList = append(vmList, tempVm)
+		}
+	}
+	t.Logf("TestVMAgent setup")
+	for _, vm := range vmList {
+		stopAntreaAgent(t, data, vm)
+		t.Logf("Creating ExternalNode for VM: %s", vm.nodeName)
+		_, err := createExternalNodeCRD(data, vm.nodeName, vm.ifName, vm.ip)
+		require.NoError(t, err, "Failed to create ExternalNode")
+		startAntreaAgent(t, data, vm)
+	}
+	return vmList, nil
+}
+
+// teardownVMAgentTest deletes ExternalNode, verifies ExternalEntity is deleted
+// and verifies uplink configuration is restored.
+func teardownVMAgentTest(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyUpLinkAfterCleanup := func(vm vmInfo) {
+		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
```

Comment:
the 30 second delay for retry looks pretty long to me. Is that intentional?

Reply:
Yes. On windows we have observed sometime it may take time. And also earlier, agent was monitoring for ExternalNode every minute.  I can change it to poll for every 10 seconds, with timeout at 1 minute.